File: random_walk_vs_brownian_motion.py
import numpy as np
import matplotlib.pyplot as plt
import os

def simulate_donsker(N=250, M=5, T=1.0, seed=123):
    """
    Simulates Scaled Random Walks and Brownian Motion to demonstrate Donsker's Theorem.
    """
    np.random.seed(seed)
    dt = T / N
    t = np.linspace(0, T, N + 1)

    plt.figure(figsize=(12, 8))

    # 1. Scaled Random Walk
    plt.subplot(2, 2, 1)
    for _ in range(M):
        xi = 2 * (np.random.rand(N) > 0.5) - 1  # Bernoulli +/- 1
        S = np.concatenate(([0], np.cumsum(xi))) / np.sqrt(N)
        plt.step(t, S, linewidth=1.1, where='mid')
    plt.title('Scaled Random Walk (Discrete)')
    plt.grid(True, alpha=0.3)
    plt.ylabel('Position')

    # 2. Brownian Motion
    plt.subplot(2, 2, 3)
    for _ in range(M):
        dW = np.sqrt(dt) * np.random.randn(N)
        W = np.concatenate(([0], np.cumsum(dW)))
        plt.plot(t, W, linewidth=1.1)
    plt.title('Brownian Motion (Continuous)')
    plt.grid(True, alpha=0.3)
    plt.xlabel('Time (t)')
    plt.ylabel('Value')

    # 3. Comparison
    plt.subplot(1, 2, 2)
    # Generate distinct samples
    xi_sample = 2 * (np.random.rand(N) > 0.5) - 1
    S_sample = np.concatenate(([0], np.cumsum(xi_sample))) / np.sqrt(N)
    
    dW_sample = np.sqrt(dt) * np.random.randn(N)
    W_sample = np.concatenate(([0], np.cumsum(dW_sample)))

    plt.step(t, S_sample, 'r', label='Scaled RW', linewidth=1.5, where='mid')
    plt.plot(t, W_sample, 'b', label='Brownian Motion', linewidth=1.5, alpha=0.7)
    plt.legend()
    plt.title('Trajectory Comparison')
    plt.grid(True, alpha=0.3)

    plt.suptitle(r"Donsker's Theorem: RW $\to$ Brownian Motion", fontsize=16)
    plt.tight_layout()
    
    # Save safely
    filename = 'donsker_convergence.pdf'
    plt.savefig(filename)
    print(f"Figure saved to {os.getcwd()}/{filename}")
    plt.show()

if __name__ == "__main__":
    simulate_donsker()

File: scripts/random_walk_vs_brownian_motion.m
%% Donsker's Invariance Principle Simulation
% Context: Bachelor Thesis (TFG) - Math
% Purpose: Visualizing the convergence of a Scaled Random Walk to Brownian Motion.

clear; clc; clf;
rng(123); % Fixed seed for reproducibility

% --- Parameters ---
N = 250;        % Number of time steps
M = 5;          % Number of sample paths to plot
T = 1;          % Time horizon
dt = T/N;       % Time increment
t = linspace(0, T, N+1);

% Setup figure
fig = figure('Name', 'Donsker Convergence Analysis', ...
             'Units', 'normalized', 'Position', [.1 .1 .8 .7]);

% --- 1. Scaled Random Walk (The Discrete Process) ---
% X_i = +/- 1 with p=0.5. Scaled by 1/sqrt(N)
subplot(2,2,1); hold on; box on; grid on
for m = 1:M
    xi = 2*(rand(N,1)>0.5) - 1;      
    S  = [0; cumsum(xi)] ./ sqrt(N); 
    stairs(t, S, 'LineWidth', 1.1)
end
xlabel('t'); ylabel('S_n (Scaled)');
title('Scaled Random Walk');

% --- 2. Brownian Motion (The Limit Process) ---
% W_t ~ N(0, t). Increments are N(0, dt)
subplot(2,2,3); hold on; box on; grid on
for m = 1:M
    dW = sqrt(dt) * randn(N,1);
    W  = [0; cumsum(dW)];
    plot(t, W, 'LineWidth', 1.1)
end
xlabel('t'); ylabel('W_t');
title('Brownian Motion Simulation');

% --- 3. Functional Convergence Comparison ---
subplot(1,2,2); hold on; box on; grid on

% Generating a single realization for direct comparison
xi_sample = 2*(rand(N,1)>0.5) - 1; 
S_sample  = [0; cumsum(xi_sample)] ./ sqrt(N);
% Note: Mathematically, we compare distribution, but visually we overlay an independent BM to show scale similarity.
W_sample  = [0; cumsum(sqrt(dt)*randn(N,1))];

stairs(t, S_sample, 'r', 'LineWidth', 1.5, 'DisplayName', 'Random Walk (n=250)')
plot(t, W_sample, 'b', 'LineWidth', 1.5, 'DisplayName', 'Brownian Motion')

legend('Location','southwest');
xlabel('t'); ylabel('Value');
title('Trajectory Comparison');
sgtitle('Donsker''s Theorem: Weak Convergence');

% --- Save Output ---
% Using relative path to ensure portability across systems
if ~exist('output', 'dir')
    mkdir('output');
end
fname = fullfile('output', 'donsker_simulation.pdf');
print(fig, fname, '-dpdf', '-painters');
fprintf('Figure saved successfully to: %s\n', fname);

